### PR TITLE
cpu/nrf52: added ADC driver implementation

### PR DESCRIPTION
**Pull request by haukepetersen**
This PR takes over from  #7515 (as discussed offline with @PeterKietzmann)

The code should be working well (don't have a board available right now). Maybe someone has an idea about the slight negative values when measuring the GND potential? I am not sure if this is a showstopper, though, might be as well fixed later on as the ADC is working fine otherwise.

### CODE REVIEW

**Comment by PeterKietzmann** [comment]
Just gave it a quick short at first. Do you know what is the reference voltage? Connecting 5V results in something about 800-900 (with 10 Bit). Additionally it appears that different channels interfere which is not necessarily caused by the driver itself...

**Comment by haukepetersen** [comment]
the ref voltage should be 3v3. Could be that 5v might break your board, no idea what its doing with them.

**Comment by haukepetersen** [comment]
but interference should not be present, seems like i have to put some more time into this...

File: cpu/nrf52/periph/adc.c [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    while (NRF_SAADC->EVENTS_STOPPED == 0) {}
+
+    /* free device */
+    done();
+
+    /* hack -> the result can be a small negative number when a AINx pin is
+     * connected via jumper wire a the board's GND pin. There seems to be a
+     * slight difference between the internal CPU GND and the board's GND
+     * voltage levels?! (observed on nrf52dk and nrf52840dk) */
+    return (result < 0) ? 0 : (int)result;
```
Comment (PeterKietzmann) [inline cpu/nrf52/periph/adc.c]: So we won't ever be able to sample negative voltages?  I would prefer to remove this hack.

**Comment by PeterKietzmann** [comment]
@haukepetersen can you check the converted values for GND and 3V3 at your site and report?

**Comment by haukepetersen** [comment]
I will, but might take some days. All my hardware is still in boxes and I only will unpack after I have worked through a paper deadline..

**Comment by dylad** [comment]
I get my hands on a NRF52DK board today so I was thinking of help you there :)

Line 0 -> VDD
Line 1 -> GND
Using wire jumper
```
ADC_LINE(1): 1023
ADC_LINE(0): 2
ADC_LINE(1): 1023
ADC_LINE(0): 2
ADC_LINE(1): 1023
ADC_LINE(0): 2
ADC_LINE(1): 1023
ADC_LINE(0): 1
ADC_LINE(1): 1023
ADC_LINE(0): 0
ADC_LINE(1): 1023
ADC_LINE(0): 1
ADC_LINE(1): 1023
ADC_LINE(0): 1
ADC_LINE(1): 1023
ADC_LINE(0): 2
ADC_LINE(1): 1023
ADC_LINE(0): 1
ADC_LINE(1): 1023
ADC_LINE(0): 1
ADC_LINE(1): 1023
```
Note : I only used ADC_NUMOF(2) because there is too much output if I kept NUMOF set to 9

**Comment by PeterKietzmann** [comment]
@haukepetersen please squash!

File: cpu/nrf52/periph/adc.c (hunk 1 above)
Comment (haukepetersen) [reply cpu/nrf52/periph/adc.c]: The thing is, that our ADC interface does not support negative values as of now. Also in the current ADC configuration, the values should never become negative...

**Comment by haukepetersen** [comment]
squashed.

File: cpu/nrf52/periph/adc.c (hunk 1 above)
Comment (PeterKietzmann) [reply cpu/nrf52/periph/adc.c]: Yes, you're completely right. However, maybe we should consider To change that at some point in time.

**Pull request merged**